rust/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+
+type HmacSha256 = Hmac<Sha256>;
+
+/// An X25519 public key.
+pub trait PublicKey {
+    /// to_bytes encodes the public key as a byte array according to RFC 7748.
+    fn to_bytes(&self) -> [u8; 32];
+}
+
+/// An X25519 private key.
+pub trait PrivateKey {
+    /// PublicKey is the type of public keys corresponding to this type.
+    type PublicKey: PublicKey;
+
+    /// dh corresponds to the x25519 function from RFC 7748, computing a shared secret from a
+    /// private key and a peer key.
+    fn dh(&self, theirs: &Self::PublicKey) -> [u8; 32];
+
+    /// public_key computes the public key corresponding to this private key.
+    fn public_key(&self) -> Self::PublicKey;
+}
+
+/// Compute a GLOME tag.
+///
+/// The message counter argument `ctr` is used to prevent replay attacks in a
+/// series of messages. If only a single message is exchanged between two key
+/// pairs, it can be set to `0u8`. The counter value is not secret, but needs
+/// to be integrity protected. Usually, this is accomplished by both parties
+/// counting sent and received messages internally.
+///
+/// The returned tag can be verified by the
+/// other party, using [verify] with its private key, our public key and the
+/// counter.
+pub fn tag<T: PrivateKey>(ours: &T, theirs: &T::PublicKey, ctr: u8, msg: &[u8]) -> [u8; 32] {
+    let key = [
+        ours.dh(theirs),
+        theirs.to_bytes(),
+        ours.public_key().to_bytes(),
+    ]
+    .concat();
+
+    HmacSha256::new_from_slice(&key)
+        .expect("HMAC can take key of any size")
+        .chain_update(&[ctr])
+        .chain_update(msg)
+        .finalize()
+        .into_bytes()
+        .into()
+}
+
+/// Verify a GLOME tag.
+///
+/// This verifies tags produced by the [tag] function. The `tag` argument can
+/// be shorter than the original 32 bytes, in which case the expected tag is
+/// shortened to match the length of the given tag. The caller needs to ensure
+/// that the tag is long enough to meet the security requirements, i.e. prevent
+/// brute-forcing.
+///
+/// The return value is `true` iff the given tag matches the expected tag.
+pub fn verify<T: PrivateKey>(
+    ours: &T,
+    theirs: &T::PublicKey,
+    ctr: u8,
+    msg: &[u8],
+    tag: &[u8],
+) -> bool {
+    let key = [
+        ours.dh(theirs),
+        ours.public_key().to_bytes(),
+        theirs.to_bytes(),
+    ]
+    .concat();
+
+    HmacSha256::new_from_slice(&key)
+        .expect("HMAC can take key of any size")
+        .chain_update(&[ctr])
+        .chain_update(msg)
+        .verify_truncated_left(tag)
+        .is_ok()
+}
+
+#[cfg(feature = "dalek")]
+pub mod dalek;
+
+#[cfg(feature = "openssl")]
+pub mod openssl;
+
+/// The [tests] module provides functions to test implementations of [PrivateKey]/[PublicKey].
+#[cfg(test)]
+pub mod tests {
+    use super::*;
+    use crate::PrivateKey;
+    use hex_literal::hex;
+
+    pub fn run_vector_1<T: PrivateKey>(load_keypair: &dyn Fn([u8; 32]) -> (T, T::PublicKey)) {
```

Comment:
Since this is just a test it's not a must and more personal preference but you can use a generic type here:

```suggestion
    pub fn run_vector_1<T, F>(load_keypair: &F)
    where
        T: PrivateKey,
        F: Fn([u8; 32]) -> (T, T::PublicKey),
    {
```
Of course the same goes for `run_vector_2`.